contrib/epee/src/net_ssl.cpp

```diff
@@ -502,8 +508,23 @@ bool ssl_options_t::handshake(boost::asio::ssl::stream<boost::asio::ip::tcp::soc
     });
   }
 
-  boost::system::error_code ec;
-  socket.handshake(type, ec);
+  auto& io_service = GET_IO_SERVICE(socket);
+  boost::asio::steady_timer deadline(io_service, timeout);
```

Comment:
Both server and client (after your initial patch) had set a timer already, so this additional one wasn't needed. Did/do you think this was better for consistency? It does ensure that the client doesn't "forget" to set a timer.

Reply:
0. Zero to no overhead (almost)
1. The current implementation is clearer and it ensures `that the client doesn't "forget" to set a timer`.
2. Without the `timeout` parameter it is not obvious that invoking `m_ssl_options.handshake(socket, type, host)` will run event loop inside and the caller actually *might* cancel async handshake.

Reply:
One thing I find frustrating with ASIO is that every OS seems to have built-in send/receive timeouts for sockets. So it should be possible to do with a bit less "work" to setup the async stuff. I haven't seen any of their examples utilize this, presumable there is some platform issues.

Reply:
In general - yes. But i think our implementation is the most correct way to handle timeouts.

Think of a malicious client intentionally replying at 1 byte per second speed during handshake. OS timeouts won't handle that. 
But our approach is immune to such an attack.

Reply:
The root is the big abstraction ASIO is providing in the handshake function. Otherwise you could adjust the timeout in a loop via an extra syscall or through `select`. But even in standard invocations of `read` that are basically direct syscalls, I've never seen an example that uses the built-in timeout techniques. Probably just confuses things; examples have consistent behavior across much of the higher-level api calls.